[ASCII-865] Config setting secret_refresh_interval for secrets component

dustmop opened the pull request:
### What does this PR do?

Add a config setting `secret_refresh_interval`, representing the number of seconds to wait before attempting to refresh secrets. When secrets are refreshed it will invoke the callback that was passed to ResolveWithCallback.

### Motivation

Refreshing secrets without restarting the agent.

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

Refreshing the secrets at regular interval requires a goroutine and locking, which was not part of the secrets compotent before.

### Describe how to test/QA your changes

Create the following script named `secret-script.py` in the working directory:

```
#!/usr/bin/env python
import datetime
import json
import sys

def main():
  content = sys.stdin.read()
  obj = json.loads(content)
  handles = obj['secrets']
  result = {}
  for h in handles:
    result[h] = {
      'value': 'decoded_%s' % h,
    }
  print(json.dumps(result))

  dt = datetime.datetime.now()
  fout = open('secret-script.log', 'a')
  fout.write('[%s] %s\n' % (dt, result))
  fout.close()

if __name__ == '__main__':
  main()
```

Set its permissions properly, to satisfy the secrets component's requirements:
```
chmod 500 secret-script.py
```

Configure the secrets backend in your datadog.yml file, including the flag to refresh secrets:
```
secret_backend_command: script-script.py 
secret_backend_arguments:
  - ENC[my_secret_arg]
secret_refresh_interval: 4
```

Tail the log file to see that the secret decoder is invoked every 4 seconds:
```
tail -f secret-script.log
```

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

dustmop committed:
Secret resolution callback only bubbles up until acknowledged

Once a resolution callback gets acknowledged, disable the `notificationPending` flag so that the callback stops being repeatedly called.

dustmop committed:
Config setting secret_refresh_interval for secrets component

Change ResolveWithCallback to RegisterResolveCallback. This method registers the configuration text, origin, and callback within the resolver, to be reused when Refresh() is called. Only one set of arguments can be registered at a time.

Using the config setting `secret_refresh_interval` allows secrets to be refreshed each given number of seconds. When refreshed, this will invoke the callback passed to the registration.

dustmop committed:
Add release notes for secret_refresh_interval

Re-enable the cache being added to by fetchSecret

GustavoCaso commented on comp/core/secrets/go.mod:
```diff
@@ -21,6 +21,7 @@ require (
 	github.com/DataDog/datadog-agent/pkg/util/winutil v0.0.0-00010101000000-000000000000
 	github.com/stretchr/testify v1.8.4
 	go.uber.org/fx v1.18.2
+	golang.org/x/exp v0.0.0-20231214170342-aacd6d4b4611
```
Is this change related to the secret refresh new code?

GustavoCaso replied:
I see we are now using `"golang.org/x/exp/slices"` 😄

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -199,17 +210,61 @@ func isEnc(str string) (bool, string) {
+	r.refreshData = data
+	r.refreshOrigin = origin
+	r.refreshCallback = cb
+	if r.refreshInterval > 0 && r.lastRefresh.IsZero() {
```
It looks like that is only the first time we've called `RegisterResolveCallback` that this gorouotine will be created, is that expected? Is `RegisterResolveCallback` meant to be called once? What would happen is someone calls this function twice with different `secrets.ResolveCallback`?

dustmop replied:
Yes, that is expected. The secretResolver is called numerous times with `.Resolve()`, but only once with `.RegisterResolveCallback()`, in `pkg/config/config.go` passing in the config textual data and notification callback. So in the current design, there is a one-to-one connection between that config data and the callback. We may revisit this in the future, so that individual components register to receive the callback, but that would require singling out the config data as being the text over which to resolve secrets. I'd rather wait for that use case to arrive than to anticipate it early. I added a check to this function to ensure its only called once per secrets object.

hush-hush replied:
I think the secret comp should be agnostic of its use. Having a method that can't be called more that once is error-prone in a context of a component.

Also we might want to resolve secrets in other configuration this way. For example, a single process might want to resolve secrets from both `datadog.yaml` and `system-probe.yaml`. I think we need a generic and future proof solution to refresh secrets.

I'm not a fan of saving the yaml internally since the data in the configuration could be updated by outside sources (remote-config, CLI, runtime-settings, ...). Since we have the source, context and the yaml path of all secrets in the `cache`, we could directly refresh them and them notify a list of callbacks. We would call the secret executable with all our secrets, refresh the cache and for each update value emit a notification with the source, the yaml path and the new value. Then it's up to the registered caller to update themselves.

That way we can refresh any secrets, not only the ones link to a single configuration and we won't need to parse again the configuration each time we want to check for new values. WDYT ?

Side note: we want to limit this to the API_KEY for now, but we should be ready to support more.

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -199,17 +210,61 @@ func isEnc(str string) (bool, string) {
+	r.refreshCallback = cb
+	if r.refreshInterval > 0 && r.lastRefresh.IsZero() {
+		r.lastRefresh = time.Now()
+		go func() {
+			for {
+				var since time.Duration
+				func() {
+					r.lock.Lock()
+					defer r.lock.Unlock()
+					since = time.Since(r.lastRefresh)
+				}()
+				if since.Seconds() > float64(r.refreshInterval) {
+					if err := r.Refresh(); err != nil {
+						log.Info(err)
+					}
+					since = 0
+				}
+				time.Sleep(time.Duration(r.refreshInterval) - since)
+			}
+		}()
+	}
```
I don't know if we should worry about this with Go, but would it make sense to stop this goroutine when the agent stops? If that is the case, we could use the fx lifecycle events for that 😄

dustmop replied:
I'm not sure if fx lifecycle would improve this, I don't have much experience with their use just yet. As is now, this routine is only sometimes started (if the `secret_refresh_interval` flag is non-zero) and will shut down fine when exiting the agent process.

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -263,6 +318,10 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
 		if r.fetchHookFunc != nil {
 			// hook used only for tests
 			secrets, err = r.fetchHookFunc(newHandles)
+			// add results to the cache
+			for handle, secretValue := range secrets {
+				r.cache[handle] = secretValue
+			}
 		} else {
 			secrets, err = r.fetchSecret(newHandles)
 		}
```
Weirdly, we have to manually handle the cache store for the test case branch. It looks like `fetchSecret` is not only fetching the secret but also updateiung the cache storage. 

I think we could improve the readbility ofthe code by having the cache by handle ins the same pace for both branches: `fetchHookFunc` , and `fetchSecret`

What about code that looks like this:
```
if r.fetchHookFunc != nil {
	// hook used only for tests
	secrets, err = r.fetchHookFunc(newHandles)
} else {
	secrets, err = r.fetchSecret(newHandles)
}

// add results to the cache
for handle, secretValue := range secrets {
	r.cache[handle] = secretValue
}

```

dustmop replied:
For sure. During development, an earlier version of the code did this, but I realized TestFetchSecret depended upon the cache getting updated inside `r.fetchSecret`. However, I think you're right that this reads much better, and there's no need to keep that test focusing on such implementation details.

GustavoCaso commented:
@dustmop With the effort to reduce QA cards, I'm wondering if, with the unit test you added, we could avoid the manual QA step for this new feature? If we can't, can we add any integration test to remove the QA process?

dustmop commented:
> @dustmop With the effort to reduce QA cards, I'm wondering if, with the unit test you added, we could avoid the manual QA step for this new feature? If we can't, can we add any integration test to remove the QA process?

I think it's good to keep a QA step to validate that the feature works from flag to behavior. I would support adding an integration test eventually, but I think it's too early just yet as this functionality may change when we develop the next steps.

GustavoCaso commented:
> I think it's good to keep a QA step to validate that the feature works from flag to behavior. I would support adding an integration test eventually, but I think it's too early just yet as this functionality may change when we develop the next steps.

Thanks for the clarification

GustavoCaso commented on comp/core/secrets/secretsimpl/fetch_secret_test.go:
```diff
@@ -249,9 +249,7 @@ func TestFetchSecret(t *testing.T) {
 	}, resp)
 	assert.Equal(t, map[string]string{
-		"test":    "yes",
-		"handle1": "p1",
-		"handle2": "p2",
```
Are these changes expected?

dustmop replied:
Yes, before `fetchSecret` was storing results in the cache. But this change is moving the cache writes up into `resolve`, but the test is directly using `fetchSecret`.

GustavoCaso replied:
Thanks for the clarification

dustmop committed:
Remove config.MergeConfigOverride

dustmop committed:
Subscribe method will notify about secret changes

Add Subscribe method, instead of ResolveWithCallback
Remove notify from walker, instead use the cache to determine what has changed
Allow assigning to Config with arbitrary paths, removing need for Resolve callback to return true/false

dustmop commented:
Thanks for the suggestions @hush-hush and @GustavoCaso, did a bit of an overhaul based upon your feedback.

The Secret component now has a `Subscribe` method for receiving updates. Its callback has this signature: `Subscribe(handle string, path []string, oldValue, newValue any)`. This gets called for each `handle` per `path` it appears at. There's no longer a need for this callback to return a bool. Instead the `Config` updates unknown fields by walking up the given path until it finds a known field, and then modifying the object at that field. See `configAssignAtPath` in pkg/config/setup/config.go and associated tests. The walker no longer has a notifier field, instead the resolver builds the origin map (linking handles to paths) and that map is used for updating subscribers during `Resolve` and `Refresh`.

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -356,6 +322,22 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+		// notify subscriptions about the changes to secrets
+		for handle, newValue := range secretResponse {
+			oldValue := r.cache[handle]
+			for _, ctx := range r.origin[handle] {
```
Since `ctx` is a common standard variable name that we tend to associate with `context.Context` why don't we rename this variable to `secretCtx`?

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -365,6 +347,45 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+		}
+		for _, sub := range r.subscriptions {
+			contexts := r.origin[handle]
+			for _, ctx := range contexts {
```
Sam as https://github.com/DataDog/datadog-agent/pull/21600/commits/7e92f9ca25400756be01c3b96b23ebf8f59c9e25#r1433796779

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -47,6 +48,14 @@ func Module() fxutil.Module {
+	origin string
+	// path is the key associated to the secret in the YAML configuration.
+	// Example: in this yaml: '{"token": "ENC[token 1]"}', 'token' is the path and 'token 1' is the handle.
+	path []string
```
We changed from a `string` to a `[]string.` Does the comment above need to be updated?

dustmop replied:
Good suggestion, updated the example to demonstrate the distinction.

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets_test.go:
```diff
@@ -85,14 +63,14 @@ instances:
-				origin:   "test",
-				yamlPath: "instances/password",
+				origin: "test",
+				path:   []string{"instances", "0", "password"},
```
The `0` and `1` changes is because of how yaml handle nested values?

dustmop replied:
Yes, these are indexes into an array of values. Before, these indexes would just be ignored, but now they're part of the path, because the new method `configAssignAtPath` is able to make use of them to modify a single element of an array.

GustavoCaso commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
+	// elements to figure out how to modify that particular object, before setting it back
+	// on the config.
+	//
+	// Example: configAssignAtPath({'process_config', 'additional_endpoints', 'http://url.com', '0'}, 'password')
```
Do the comment miss the config object as the first argument?

dustmop replied:
Yes, it was missed, added now.

pgimalac commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -196,20 +205,44 @@ func isEnc(str string) (bool, string) {
+// Subscribe adds this callback to the list that get notified when secrets are resolved or refreshed
+func (r *secretResolver) Subscribe(cb secrets.ResolveCallback) {
+	r.startRefreshRoutine()
+	r.subscriptions = append(r.subscriptions, cb)
```
I think there is a race here, you write to `r.subscriptions` without locking, it can be read concurrently ?

GustavoCaso replied:
Nice catch 🎉

dustmop replied:
Sure that makes sense, moved the lock from `startRefreshRoutine` into here.

GustavoCaso commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
-				return true
-			})
-		if err != nil {
+		secretResolver.Subscribe(func(handle string, settingPath []string, oldValue, newValue any) {
+			if err := configAssignAtPath(config, settingPath, newValue); err != nil {
+				log.Errorf("could not assign to config: %s", err)
+			}
+		})
+		if _, err = secretResolver.Resolve(yamlConf, origin); err != nil {
```
Looking at the secretResolver functions, I find the names a bit cryptic.

`Subscribe` subscribe to what? When is the subscribe callback going to get called? What about changing the name to `RegisterResolveCallback` I think it express what do actual secret resolvers doing 😄

dustmop replied:
I'm not a fan of `RegisterResolveCallback` (despite using it in an earlier version) because it's talking about a type that is usually not going to appear in code, since most users will be anonymous functions. I also think `ResolveCallback` needs an update because it's only mentioning `Resolve` but now secrets can be updated with `Refresh` as well. Going to go with `SecretChangeCallback` and `SubscribeToChanges` which I think together are a pretty good solution to all these issues.

GustavoCaso commented:
@dustmop, great work 🎉 

I like that we were able to simplify the walker interface. I still find the code in `configAssignAtPath` a bit cryptic. I know it is because of Viper limitations 😄 Still, I don't have enough expertise with it that I will let @hush-hush add the last stamp of approval.

One question: Is this code going to be able to be used by all the configurations?

pgimalac commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +347,45 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+		return err
+	}
+
+	for handle, secretValue := range secretResponse {
+		prevValue := r.cache[handle]
+		if prevValue == secretValue {
+			continue
+		}
+		for _, sub := range r.subscriptions {
+			contexts := r.origin[handle]
+			for _, ctx := range contexts {
+				sub(handle, ctx.path, prevValue, secretValue)
+			}
+		}
+		r.cache[handle] = secretValue
+	}
```
This block is (mostly) duplicated above, it might make sense to move it to an helper function

dustmop replied:
Done, now both call sites are using `handleSecretResponse`.

pgimalac commented on pkg/config/config_template.yaml:
```diff
@@ -839,6 +839,12 @@ api_key:
 
+## @param secret_refresh_interval - float - optional - default: 0
+## @env DD_SECRET_REFRESH_INTERVAL - float - optional - default: 0
+## Interval at which to refresh the secrets
```
Can you specify the default behavior (disabled) when left to 0 ?

pgimalac commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
 	return nil
 }
 
+func configAssignAtPath(config pkgconfigmodel.Config, settingPath []string, newValue any) error {
```
Could you please add a small (or not) doc comment to explain what this function does ?

pgimalac commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
+
+	trailingElements := []string{}
+	// copy the path and hold onto the original, useful for error messages
+	path := make([]string, len(settingPath))
+	copy(path, settingPath)
```
Could use `slices.Clone` here (`path := slices.Clone(settingPath)`)

dustmop replied:
Nice, I had searched for this but missed it in the `slices` package.

pgimalac commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1767,104 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
+		}
+		// get the last element from the path and prepend it to the trailing elements
+		lastElem := path[len(path)-1]
+		trailingElements = append([]string{lastElem}, trailingElements...)
```
Instead of repetitively prepending, you can `append` and `slices.Reverse` at the end of the loop

pgimalac replied:
Slices are really arrays (https://go.dev/blog/slices-intro) so prepending means allocating a new array each time and copying all elements of the previous slice.
If you `append` instead in the loop, and just `Reverse` it once the loop is done, at best (if the capacity was enough) you only copy elements once when reversing it.
That's n allocs + n^2 copy vs 1 alloc + n copy

GustavoCaso replied:
Thanks for the explanation @pgimalac 

> at best (if the capacity was enough)

Would that means that we need to specify the len parameter when creating the initial slice? 

```go
trailingElements := make([]string, len(settingName))
```

That way, we guarantee that calling `append` in the loop will not allocate extra space.

dustmop replied:
Almost, the most efficient code is to not specify the length but the capacity, using the third argument:

```
trailingElements := make([]string, 0, len(settingPath))
```

GustavoCaso replied:
Almost. I'm still, getting used to the slice internal in Go 😄 

Thanks for the feedback

dustmop committed:
Various fixes for review comments, comments and cleanups

dustmop commented on go.sum:
I don't know what happened there, but I redid it, should be better now.

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -196,20 +205,44 @@ func isEnc(str string) (bool, string) {
+					since = 0
+				}
+				time.Sleep(time.Duration(r.refreshInterval) - since)
+			}
```
Ah that's much better!

dustmop committed:
Rename Subscribe to SubscribeToChanges and ResolveCallback to SecretChangeCallback

dustmop commented:
> One question: Is this code going to be able to be used by all the configurations?

Any configuration that's resolving secrets will be using this code, but will only be affected by these changes if they're subscribing to updates. The main client code is in `pkg/config/setup/config.go` which is loading the main `datadog.yml` config and most needs this granular form of secret updating in order to preserve source information (where values are coming from) in the config. Other callers will often use secrets.Resolve's return value instead, but doing that for `datadog.yml` will erase the source info.

dustmop committed:
Fix tests

dustmop committed:
Whitelist for SecretsResolver, only api_key can be Refreshed

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +327,59 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+// tests can override this to exercise functionality
+var whitelistHandles = []string{"api_key"}
+
+func (r *secretResolver) processSecretResponse(secretResponse map[string]string, bypassWhitelist bool) {
+	// notify subscriptions about the changes to secrets
+	for handle, secretValue := range secretResponse {
+		// if whitelist is enabled and the handle is not contained in it, skip it
+		if whitelistHandles != nil && !bypassWhitelist && !slices.Contains(whitelistHandles, handle) {
+			continue
+		}
+		oldValue := r.cache[handle]
+		// if value hasn't changed, don't send notifications
+		if oldValue == secretValue {
+			continue
+		}
+		for _, secretCtx := range r.origin[handle] {
+			for _, sub := range r.subscriptions {
+				sub(handle, secretCtx.path, oldValue, secretValue)
+			}
+		}
+		// add results to the cache
+		r.cache[handle] = secretValue
+	}
+}
```
```suggestion
func (r *secretResolver) processSecretResponse(secretResponse map[string]string, bypassWhitelist bool) {
	process := func(handle, secretValue string) {
		oldValue := r.cache[handle]
		// if value hasn't changed, don't send notifications
		if oldValue == secretValue {
			return
		}
		for _, secretCtx := range r.origin[handle] {
			for _, sub := range r.subscriptions {
				sub(handle, secretCtx.path, oldValue, secretValue)
			}
		}
		// add results to the cache
		r.cache[handle] = secretValue
	}

	// notify subscriptions about the changes to secrets
	for handle, secretValue := range secretResponse {
		if bypassWhitelist {
			process(handle, secretValue)
		} else {
			// if whitelist is enabled and the handle is contained in it, process it it
			if whitelistHandles != nil && slices.Contains(whitelistHandles, handle) {
				process(handle, secretValue)
			}
		}
	}
}
```

I find the condition `whitelistHandles != nil && !bypassWhitelist && !slices.Contains(whitelistHandles, handle)` very hard to grok 😄 

If we create a separate inline function `process` we can have a cleaner branching logic 😄 

```go
if bypassWhitelist {
	process(handle, secretValue)
} else {
	// if whitelist is enabled and the handle is contained in it, process it it
	if whitelistHandles != nil && slices.Contains(whitelistHandles, handle) {
		process(handle, secretValue)
	}
}
```

dustmop replied:
I think if I saw that code I would notice the duplication of `process(handle, secretValue)` and want to merge it back together. But I see your point about the condition being strange to understand. I think it may be better if the bool is inverted and moved to the front of the condition, which reads a bit more logically:

```
if useWhitelist && whitelistHandles != nil && !slices.Contains(whitelistHandles, handle) {
```

What do you think?

GustavoCaso replied:
Yeah 🎉 

That condition is much more readable 😄

GustavoCaso commented on comp/core/secrets/type.go:
```diff
@@ -11,40 +11,9 @@ type SecretVal struct {
-type ResolveCallback func(key []string, value any) bool
+// SecretChangeCallback is the callback type used by SubscribeToChanges to send notifications
+// This callback will be called once for each time a handle at a particular path is resolved or refreshed
+type SecretChangeCallback func(handle string, path []string, oldValue, newValue any)
```
What about `SecretChangedCallback`?

dustmop replied:
No, I prefer to avoid past tense here. I think of the callback as something that is "currently happening" not something that has finished already and we are responding to, regardless of the implementation details around it. If you look around the agent codebase for other callbacks, there's a similar naming convention: `APITokenRenewCallback`, `scrubCallback`, `onDeleteCallback`...

GustavoCaso replied:
Thanks for the clarification 👍

GustavoCaso commented on comp/core/secrets/secretsimpl/secrets_test.go:
```diff
@@ -366,30 +367,105 @@ func TestResolveWithCallback(t *testing.T) {
+	testConf := testConfNestedMultiple
+
+	// disable the whitelist for the test, let any secret changes happen
+	originalWhitelistHandles := whitelistHandles
+	whitelistHandles = nil
+	defer func() { whitelistHandles = originalWhitelistHandles }()
```
I'm not sure if you added it already, but should we add a test that exercises `whitelistHandles` having a value and only subscription being called for that value and not others?

GustavoCaso commented on comp/core/secrets/secretsimpl/walker.go:
```diff
@@ -7,82 +7,51 @@ package secretsimpl
 )
 
-type resolverCallback func([]string, string) (string, error)
+type resolveFunc func([]string, string) (string, error)
```
Do we still need this type?

dustmop replied:
Nope, good catch, it can be removed.

GustavoCaso replied:
I mean that we might not need a type definition if is not public.

```go
type walker struct {
	// resolver is called to fetch the value of a handle
	resolver func([]string, string) (string, error)
}
```

dustmop replied:
Ah whoops, yes of course.

dustmop committed:
Invert bool bypassWhitelist, change it to useWhitelist

dustmop committed:
Test for whitelist of secretResolver.Refresh

GustavoCaso reviewed (approved):
Amazing work @dustmop 🎉

dustmop committed:
Inline definition of resolverFunc

pgimalac commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +327,59 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	newHandles := make([]string, 0, len(r.cache))
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
**❓ question:** ‏Do we actually need to clone it ?

dustmop replied:
Yes, need to convert the map[string]... into a []string. Using `maps.Keys` for that, thanks for the tip!

pgimalac replied:
Ah yes I wrote that before realizing that it was a `map` :smile:

pgimalac commented on pkg/config/model/viper.go:
```diff
@@ -526,13 +526,6 @@ func (c *safeConfig) MergeConfig(in io.Reader) error {
 	return c.Viper.MergeConfig(in)
 }
 
-// MergeConfigOverride wraps Viper for concurrent access
-func (c *safeConfig) MergeConfigOverride(in io.Reader) error {
-	c.Lock()
-	defer c.Unlock()
-	return c.Viper.MergeConfigOverride(in)
-}
-
```
Any particular reason to remove this function ?

dustmop replied:
The only caller to this method was removed by https://github.com/DataDog/datadog-agent/pull/21223, and this PR is a follow-up to that, building on changes it made. There's undesirable behavior in `MergeConfigOverride`, namely that it removes source information from the config, that this feature work is improving by making config modifications more fine-grained.

pgimalac replied:
If it's on purpose then perfect for me :)

pgimalac reviewed (approved):
A few more nits, but LGTM

pgimalac commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +327,59 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	newHandles := make([]string, 0, len(r.cache))
+	for handle := range r.cache {
+		newHandles = append(newHandles, handle)
+	}
```
Ah wait `r.cache` is a map, so it should be `maps.Keys(r.cache)` instead
Note that the `Keys` function of `exp/maps` did not make it in the non-experimental version of `maps`, so we might not want to start using it

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -196,20 +206,36 @@ func isEnc(str string) (bool, string) {
+	if r.ticker != nil || r.refreshInterval == 0 {
+		return
+	}
+	r.ticker = time.NewTicker(time.Duration(r.refreshInterval) * time.Second)
```
The cast is unfortunately needed. If the left-hand-side were a literal constant (such as `4 * time.Second`) then no cast is required, but given that `r.refreshInterval` is an int, it needs to be cast to `time.Duration` here.

pgimalac replied:
Aaaah `time.Duration` is `int64` ! You are completely right thanks !

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +327,59 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
 	return finalConfig, nil
 }
 
+// whitelistHandles restricts what config settings may be updated
+// tests can override this to exercise functionality
+var whitelistHandles = []string{"api_key"}
```
I prefer it stay close to where it's being used in this case, for a few reasons:
1) This data is only used for refreshing secrets, not for the larger secrets functionality. If it were at the top that may confuse a reader into thinking it's more important / more generally applicable to the secrets feature.
2) This is only used in a single place in the file, and needs to be top-level only so that tests can override it. If it were used multiple places in the file it would make more sense at the top.
3) This is theoretically only needed temporarily, for the initial phase of the api_key refresh feature. Long-term it may not be needed, once arbitrary parts of configuration can be refreshed.

After thinking about this more, I also moved `secretInfoTmpl` to be closer to the single place where it is used. This used to be closer but the component conversion moved it away.

The point about `authorizedConfigPathsCore` is a good one. Added a comment linking it to that variable to hopefully prevent bugs in the future. I don't think these two values should be merged just yet. Each is meant for specific features, that may be closely aligned, but aren't necessarily the same thing.

pgimalac replied:
Fair points about the position of the variable :+1:

pgimalac replied:
> I don't think these two values should be merged just yet. Each is meant for specific features, that may be closely aligned, but aren't necessarily the same thing.

Yes I agree, we can consider later if they should always be the same set of paths or not

dustmop committed:
Tweaks and cleanups for secrets

hush-hush commented on releasenotes/notes/refresh-secrets-setting-1865ad684a8c27d0.yaml:
```diff
@@ -0,0 +1,12 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+features:
+  - |
+    Config setting `secret_refresh_interval` will periodically refresh
+    secrets at the given interval.
```
Same here: as the feature is not yet usable we should avoid documenting it for customers.

hush-hush commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1768,108 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
 	return nil
 }
 
+// assign a value to the given setting of the config
+// This works around viper issues that prevent us from assigning to fields that have a dot in the
+// name (example: 'additional_endpoints.http://url.com') and also allows us to assign to individual
+// elements of a slice of items (example: 'proxy.no_proxy.0' to assign index 0 of 'no_proxy')
```
```suggestion
// configAssignAtPath assigns a value to the given setting of the config
// This works around viper issues that prevent us from assigning to fields that have a dot in the
// name (example: 'additional_endpoints.http://url.com') and also allows us to assign to individual
// elements of a slice of items (example: 'proxy.no_proxy.0' to assign index 0 of 'no_proxy')
```

hush-hush commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1768,108 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
+	// elements to figure out how to modify that particular object, before setting it back
+	// on the config.
+	//
+	// Example: configAssignAtPath(config, ['process_config', 'additional_endpoints', 'http://url.com', '0'], 'password')
+	//
```
```suggestion
	// Example with the following configuration:
	//
	//    process_config:
	//      additional_endpoints:
	//        http://url.com:
	//         - ENC[handle to the password]
	//
	// this function willl be called with:
	//    configAssignAtPath(config, ['process_config', 'additional_endpoints', 'http://url.com', '0'], 'password')
	//
```
WDYT ?

dustmop replied:
Definitely helps explain it better, thanks!

hush-hush commented on pkg/config/setup/config_test.go:
```diff
@@ -1100,3 +1099,72 @@ func TestProxyLoadedFromConfigFileAndEnvVars(t *testing.T) {
 }
+
+var testExampleConf = []byte(`
+secret_backend_command: some command
+additional_endpoints:
+  https://url1.com:
+    - ENC[api_key_1]
+    - ENC[api_key_2]
+  https://url2.eu:
+    - ENC[api_key_3]
+process_config:
+  additional_endpoints:
+    https://url1.com:
+      - ENC[api_key_1]
+      - ENC[api_key_2]
+    https://url2.eu:
+      - ENC[api_key_3]
+`)
```
Could we remove the use of `ENC[]` ? When seeing the test case I first thought we were testing secrets replacement.

dustmop replied:
Good idea, since this is only testing `configAssignAtPath`. Replaced the example config text.

dustmop committed:
Use allowlist and update some tests

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -229,22 +253,21 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
 
 	w := &walker{
-		resolver: func(yamlPath []string, value string) (string, error) {
+		resolver: func(path []string, value string) (string, error) {
```
I think I'm misunderstanding a use case here. I know there's specific needs for the Config: it wants to get callbacks with fine-grained information about what secrets are resolved the first time. Right now this is handled by Resolve() ignoring the allowlist and sending every change. Is this ever going to happen while the cache has values in it? Is that what you mean by "used in a new configuration"?

dustmop replied:
Okay, added code here to update subscriptions when a cached value is found. I think it makes sense: the cache is for efficiency (pertaining to what gets fetched from the backend command) but shouldn't affect subscriptions as far as `Resolve` is concerned, since subscribers need to know about the initial resolution state.

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +325,57 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+// whitelistHandles restricts what config settings may be updated
+// tests can override this to exercise functionality
+// NOTE: Related feature to `authorizedConfigPathsCore` in `comp/api/api/apiimpl/internal/config/endpoint.go`
+var whitelistHandles = []string{"api_key"}
```
Yes, much better, change it

dustmop commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +325,57 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+	// notify subscriptions about the changes to secrets
+	for handle, secretValue := range secretResponse {
+		// if whitelist is enabled and the handle is not contained in it, skip it
+		if useWhitelist && whitelistHandles != nil && !slices.Contains(whitelistHandles, handle) {
```
The check for `allowlistHandles != nil` means tests can set `allowlistHandles = nil` to disable the list and allow everything. Just added a comment mentioning this.

dustmop commented on comp/core/secrets/secretsimpl/walker.go:
```diff
@@ -7,82 +7,49 @@ package secretsimpl
+
+		path := make([]string, len(yamlPath), len(yamlPath)+1)
+		copy(path, yamlPath)
+		path = append(path, strconv.Itoa(idx))
```
No, this functionality still works, even if a map has keys that look like the ints. The reason is that `configAssignAtPath` does not check the path to see if part of it "looks like" an int, rather it type checks the receiving object, and only converts the path piece into an int if the receiving object is a slice. Added a test case that demonstrates it working correctly.

hush-hush replied:
I see, thanks for the explanation. Could we document that in the docstring for the `Register` method of the `secret` comp so a caller knows exactly how to act ?

dustmop committed:
Update documentation

dustmop committed:
Notify subscribers even if value is cached

hush-hush commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -61,6 +75,11 @@ type secretResolver struct {
 	// responseMaxSize defines max size of the JSON output from a secrets reader backend
 	responseMaxSize int
+	// refresh secrets at a regular interval
+	refreshInterval int
```
We should use a `time.Duration` here to avoid negative values

hush-hush commented on comp/core/secrets/secretsimpl/secrets.go:
```diff
@@ -299,6 +330,70 @@ func (r *secretResolver) resolve(data []byte, origin string, notifyCb secrets.Re
+		}
+		oldValue := r.cache[handle]
+		// if value hasn't changed, don't send notifications
+		if oldValue == secretValue {
+			continue
+		}
```
If a handle is used multiple times in the same configuration we would only notify on the first one.

dustmop replied:
Fixed by moving the assignment to the cache to a second loop.

hush-hush commented on comp/core/secrets/type.go:
```diff
@@ -11,40 +11,9 @@ type SecretVal struct {
-// "process_config.additional_endpoints.https://app.datadoghq.com" but accept the one for the parent
-// "process_config.additional_endpoints".
-type ResolveCallback func(key []string, value any) bool
+// SecretChangeCallback is the callback type used by SubscribeToChanges to send notifications
+// This callback will be called once for each time a handle at a particular path is resolved or refreshed
```
Like I said in my other comment, we should explain how the callback should behave and what the parameters it receive will be.

dustmop replied:
Done, added comments to each argument.

hush-hush commented on pkg/config/setup/config.go:
```diff
@@ -1765,38 +1768,117 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
-			})
-		if err != nil {
+		secretResolver.SubscribeToChanges(func(handle, origin string, settingPath []string, oldValue, newValue any) {
+			if err := configAssignAtPath(config, settingPath, newValue); err != nil {
```
Le'ts check the `origin` too.

dustmop committed:
Switch refreshInterval to time.Duration, add comments

dustmop committed:
Fix lint

hush-hush reviewed (approved):
Looks good :+1:  !

GustavoCaso commented on pkg/config/setup/config.go:
```diff
@@ -1768,7 +1768,10 @@ func ResolveSecrets(config pkgconfigmodel.Config, secretResolver secrets.Compone
 
-		secretResolver.SubscribeToChanges(func(handle, origin string, settingPath []string, oldValue, newValue any) {
+		secretResolver.SubscribeToChanges(func(handle, settingOrigin string, settingPath []string, oldValue, newValue any) {
+			if origin != settingOrigin {
+				return
+			}
```
Should we add a test case for this branch?  😄

dustmop replied:
Surely, test added.

GustavoCaso commented on pkg/config/setup/config_test.go:
```diff
@@ -1100,3 +1099,108 @@ func TestProxyLoadedFromConfigFileAndEnvVars(t *testing.T) {
+}
+
+func TestConfigAssignAtPathForIntMapKeys(t *testing.T) {
+	// CircleCI sets NO_PROXY, so unset it for this test
+	unsetEnvForTest(t, "NO_PROXY")
```
Could we explain why ENV var modifies the behaviour of the test?

dustmop replied:
I have to admit, I mostly just copy and pasted this code from other tests, and so don't totally understand it. Without this line, the proxy section of the config is filled out, but the details within depend upon the current machine, so the test can't be written in a way that passes on CI. I think it would be good to fix this situation, but that may be out of scope for this PR.

GustavoCaso replied:
Thanks for the explanation. 

> I think it would be good to fix this situation, but that may be out of scope for this PR.

I totally agree with that 😄

dustmop committed:
Additional tests for secret resolution, fix some edge cases

GustavoCaso reviewed (approved):
🚀
